ludwig/models/retrieval.py

```diff
@@ -0,0 +1,225 @@
+from ludwig.utils.torch_utils import get_torch_device
+
+
+def df_checksum(df: pd.DataFrame) -> str:
+    return hashlib.sha1(pd.util.hash_pandas_object(df).values).hexdigest()
+
+
+class RetrievalModel:
+    def create_dataset_index(self, df: pd.DataFrame, backend: "Backend", columns_to_index: Optional[List[str]] = None):
+        """Creates an index for the dataset.
+
+        If `columns_to_index` is None, all columns are indexed. Otherwise, only the columns in `columns_to_index` are
+        used for indexing, but all columns in `df` are returned in the search results.
+        """
+        raise NotImplementedError
+
+    def search(
+        self, df, backend: "Backend", k: int = 10, return_data: bool = False
+    ) -> Union[List[int], List[Dict[str, Any]]]:
+        """Retrieve the top k results for the given query.
+
+        If `return_data` is True, returns the data associated with the indices. Otherwise, returns the indices.
+        """
+        raise NotImplementedError
+
+    def save_index(self, name: str, cache_directory: str):
+        """Saves the index to the cache directory."""
+        raise NotImplementedError
+
+    def load_index(self, name: str, cache_directory: str):
+        """Loads the index from the cache directory."""
+        raise NotImplementedError
+
+
+class RandomRetrieval(RetrievalModel):
+    """Random retrieval model.
+
+    Gets k random indices from the dataset regardless of the query.
+    """
+
+    def __init__(self, **kwargs):
+        self.index = None
+        self.index_data = None
+
+    def create_dataset_index(self, df: pd.DataFrame, backend: "Backend", columns_to_index: Optional[List[str]] = None):
+        self.index = np.array(range(len(df)))
+        self.index_data = df
+
+    def search(
+        self, df, backend: "Backend", k: int = 10, return_data: bool = False
+    ) -> Union[List[int], List[Dict[str, Any]]]:
+
+        results = []
+        for _ in tqdm(range(len(df))):
+            indices = np.random.choice(self.index, k, replace=False)
+
+            if return_data:
+                result = self.index_data.iloc[indices].to_dict(orient="records")
+            else:
+                result = indices
+            results.append(result)
+        return results
+
+    def save_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        # open file to prevent using the .npy extension
+        # https://numpy.org/doc/stable/reference/generated/numpy.save.html
+        with open(index_file_path, "wb") as f:
+            np.save(f, self.index)
+
+        index_data_file_path = os.path.join(cache_directory, name + "_data.csv")
+        self.index_data.to_csv(index_data_file_path, index=False)
+
+    def load_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        self.index = np.load(index_file_path)
```

Comment:
It might be worth wrapping this in a try-except block incase there is no index at the cache path

Reply:
I think we do want it to fail with the default `FileNotFoundError` if there is no index at the cache path, so I'm leaning towards leaving it as-is. Is there a reason we might want to do otherwise?

Reply:
That makes sense yeah, we can leave it as is!